source/iopipe/json/formatter.d

```diff
@@ -0,0 +1,466 @@
+        putIndent();
+        putStr("}");
+
+        popContainer(); // State.End or State.Comma
+    }
+
+    void beginArray() {
+        final switch (state) with (State)
+        {
+        case Begin, Member, Value:
+            break;
+        case First, End:
+            throw new JSONIopipeException(
+                "beginArray not allowed in current state");
+        }
+
+        pushContainer(false); // array
+        state = State.First;
+
+        ++spacing;
+        putStr("[");
+        putIndent();
+    }
+
+    void endArray() {
+        enforce(!inObj(), "endArray outside array");
+
+        --spacing;
+        putIndent();
+        putStr("]");
+
+        popContainer();
+    }
+
+    // Public hook for arrays: call before each element
+    void beginArrayValue()
+    {
+        startArrayValue();
+    }
+
+    void addMember(T)(T key) {
+        addMember(key, memberOptions);
+    }
+
+    void addMember(T)(T key, MemberOptions options) {
+        
+        enforce(inObj(), "addMember only valid inside object");
+
+        final switch (state) with (State)
+        {
+        case First, Member:
+            break;
+        case Begin, Value, End:
+            throw new JSONIopipeException(
+                "addMember not allowed in current state");
+        }
+
+        startObjectMember();
+
+        static if (is(typeof(key) == string))
+        {
+            if (options.escapeKey)
+            {
+                beginString(options.quoteChar);
+                addStringData(key);
+                endStringRaw();
+            }
+            else
+            {
+                beginString(options.quoteChar);
+                addStringData!(false, false)(key);
+                endStringRaw();
+            }
+        }
+        else
+        {
+            putStr(to!string(key));
+        }
+
+        final switch (options.colonSpacing)
+        {
+        case MemberOptions.ColonSpacing.none:
+            putStr(":");
+            break;
+        case MemberOptions.ColonSpacing.after:
+            putStr(": ");
+            break;
+        case MemberOptions.ColonSpacing.both:
+            putStr(" : ");
+            break;
+        }
+
+        state = State.Value;
+    }
+
+    void beginString(char quoteChar = '"') {
+        // value allowed at: Begin (root), Member (array element), Value (object value)
+        final switch (state) with (State)
+        {
+        case Begin, Member, Value:
+            break;
+        case First, End:
+            throw new JSONIopipeException(
+                "beginString not allowed in current state");
+        }
+
+        if (quoteChar != '"' && quoteChar != '\'')
+            throw new JSONIopipeException("Invalid quote character");
+
+        currentQuoteChar = quoteChar;
+        putStr((&quoteChar)[0 .. 1]);
+    }
+
+    // will automatically escape string data as needed.
+    void addStringData(bool validate = true, bool addEscapes = true, T)(T value) {
```

Comment:
Need a string state to ensure we aren't just adding string data anywhere. Maybe if `currentQuoteChar` is `char.init`, it's invalid?